src/gt4py/next/type_system/type_specifications.py

```diff
@@ -92,31 +79,37 @@ def __str__(self) -> str:
         return f"{kind_str}{self.shape}"
 
 
-@dataclass(frozen=True)
+class ListType(DataType):
+    element_type: DataType
+    # TODO(havogt): the `offset_type` is not yet used in type_inference,
+    # it is meant to describe the neighborhood (via the local dimension)
+    offset_type: Optional[common.Dimension] = None
+
+
+class FieldType(DataType, CallableType):
+    dims: list[common.Dimension]
+    dtype: ScalarType | ListType
+
+    def __str__(self) -> str:
+        dims = "..." if self.dims is Ellipsis else f"[{', '.join(dim.value for dim in self.dims)}]"
+        return f"Field[{dims}, {self.dtype}]"
+
+
 class TupleType(DataType):
-    types: list[DataType]
+    # TODO: or DimensionType is a DataType
+    types: list[DataType | DimensionType | DeferredType]
```

Comment:
```suggestion
    # TODO(tehrengruber): Remove `DeferredType` again. This was erroneously
    #  introduced before we checked the annotations at runtime. All attributes of 
    #  a type that are types themselves must be concrete.
    types: list[DataType | DimensionType | DeferredType]
```
The `DataType | DimensionType` is somewhat ugly, what is actually meant here is `TypeSpec` that is not `DeferredType`, i.e. a concrete type (see `is_concrete`). Promoting `is_concrete` to something that could be expressed in an annotation, e.g. by inheriting from a new class or mixin `ConcreteType`, would be an idea.